packages/signature-v4/src/SignatureV4.ts

```diff
@@ -111,26 +111,21 @@ export class SignatureV4 implements RequestPresigner, RequestSigner, StringSigne
     this.uriEscapePath = uriEscapePath;
     // default to true if applyChecksum isn't set
     this.applyChecksum = typeof applyChecksum === "boolean" ? applyChecksum : true;
-
-    if (typeof region === "string") {
-      const promisified = Promise.resolve(region);
-      this.regionProvider = () => promisified;
-    } else {
-      this.regionProvider = region;
-    }
-
-    if (typeof credentials === "object") {
-      const promisified = Promise.resolve(credentials);
-      this.credentialProvider = () => promisified;
-    } else {
-      this.credentialProvider = credentials;
-    }
+    this.regionProvider = normalizeRegionProvider(region);
+    this.credentialProvider = normalizeCredentialsProvider(credentials);
   }
 
   public async presign(originalRequest: HttpRequest, options: RequestPresigningArguments = {}): Promise<HttpRequest> {
-    const [region, credentials] = await Promise.all([this.regionProvider(), this.credentialProvider()]);
-
-    const { signingDate = new Date(), expiresIn = 3600, unsignableHeaders, signableHeaders } = options;
+    const {
+      signingDate = new Date(),
+      expiresIn = 3600,
+      unsignableHeaders,
+      signableHeaders,
+      signingRegion,
+      signingService,
+    } = options;
+    const credentials = await this.credentialProvider();
+    const region = signingRegion || (await this.regionProvider());
```

Comment:
Use [Nullish coalescing operator (??)](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Nullish_coalescing_operator)?

```suggestion
    const region = signingRegion ?? (await this.regionProvider());
```

Ditto on other places when `||` is used: 137, 178, 180 etc